File: benchmarks/cpp/src/conformance.cpp
// Proves the C++ reference computes the same thing as the Rust engine.
//
// Run before believing any timing. `NEXORA TECHNOLOGY BENCHMARK PLAN.md` rule 1
// requires the same logical workload; two stacks are far easier to drift apart
// than two functions in one language, and a comparison between a correct
// implementation and a subtly wrong one is worse than no comparison at all.
//
// Emits `name<TAB>0x…` lines. `nexora-benchmark --conformance` emits the same
// names, and scripts/compare-stacks.sh fails on any difference.

#include <cinttypes>
#include <cstdio>
#include <string>
#include <vector>

#include "kernels.hpp"

namespace {

using nexora::Fnv1a64;

void emit(const char* name, std::uint64_t digest) {
  std::printf("%s\t0x%016" PRIx64 "\n", name, digest);
}

std::uint64_t digest_splitmix64() {
  nexora::Rng rng(0x0BEEF0BEEFULL);
  Fnv1a64 hasher;
  for (int i = 0; i < 64; ++i) {
    hasher.write_u64(rng.next_u64());
  }
  return hasher.finish();
}

std::uint64_t digest_next_below() {
  nexora::Rng rng(12345);
  Fnv1a64 hasher;
  for (int i = 0; i < 2000; ++i) {
    hasher.write_u64(rng.next_below(25));
    hasher.write_u64(rng.next_below(1));
    hasher.write_u64(rng.next_below(0));
    hasher.write_u64(rng.next_below(0xFFFF'FFFF'FFFF'FFFFULL));
  }
  return hasher.finish();
}

std::uint64_t digest_stream_seeds() {
  Fnv1a64 hasher;
  const char* streams[] = {"terrain", "biome",        "cave", "vegetation", "weather",
                           "structure", "civilization", "ai",   "history",    "loot"};
  for (std::uint64_t seed : {0ULL, 1ULL, 0xDEADBEEFULL, 0xFFFFFFFFFFFFFFFFULL}) {
    for (const char* stream : streams) {
      hasher.write_u64(nexora::derive_stream_seed(seed, stream));
    }
  }
  return hasher.finish();
}

std::uint64_t digest_positional_rng() {
  Fnv1a64 hasher;
  nexora::Rng driver(0xA5A5A5A5ULL);
  for (int i = 0; i < 512; ++i) {
    const auto x = static_cast<std::int64_t>(driver.next_u64()) % 100000;
    const auto y = static_cast<std::int64_t>(driver.next_u64()) % 1000;
    const auto z = static_cast<std::int64_t>(driver.next_u64()) % 100000;
    hasher.write_u64(nexora::positional_rng(0x5EED, "terrain", x, y, z).next_u64());
  }
  return hasher.finish();
}

std::uint64_t digest_fnv1a() {
  Fnv1a64 hasher;
  const char* inputs[] = {"", "a", "foobar", "nexora:block/stone", "the quick brown fox"};
  for (const char* input : inputs) {
    hasher.write_u64(nexora::fnv1a64(input, std::char_traits<char>::length(input)));
  }
  // And the length-prefixed string path, which is what the seed derivation uses.
  Fnv1a64 prefixed;
  prefixed.write_str("ab");
  prefixed.write_str("c");
  hasher.write_u64(prefixed.finish());
  return hasher.finish();
}

std::uint64_t digest_crc32() {
  Fnv1a64 hasher;
  const char* inputs[] = {"", "123456789", "NEXORA", "nexora:save/world_header"};
  for (const char* input : inputs) {
    hasher.write_u32(nexora::crc32(input, std::char_traits<char>::length(input)));
  }
  // A longer buffer, so the loop is exercised rather than just the short cases.
  std::vector<std::uint8_t> payload(4096);
  nexora::Rng driver(7);
  for (auto& byte : payload) {
    byte = static_cast<std::uint8_t>(driver.next_u64());
  }
  hasher.write_u32(nexora::crc32(payload.data(), payload.size()));
  return hasher.finish();
}

std::uint64_t digest_section_of() {
  const nexora::ChunkShape shape;
  Fnv1a64 hasher;
  nexora::Rng driver(0x2545F4914F6CDD1DULL);
  for (int i = 0; i < 20000; ++i) {
    const auto spread = [&driver]() {
      return static_cast<std::int64_t>(driver.next_u64()) % 4000000;
    };
    const std::int64_t x = spread();
    const std::int64_t y = spread();
    const std::int64_t z = spread();
    const nexora::SectionCoord coord = nexora::section_of(shape, x, y, z);
    hasher.write_i64(coord.x);
    hasher.write_i64(coord.y);
    hasher.write_i64(coord.z);
  }
  // The negative boundaries explicitly: truncating division mirrors the world
  // across the origin, and this is where that would show.
  for (std::int64_t value : {-33, -32, -31, -1, 0, 1, 31, 32, 33}) {
    const nexora::SectionCoord coord = nexora::section_of(shape, value, value, value);
    hasher.write_i64(coord.x);
  }
  return hasher.finish();
}

std::uint64_t digest_surface_height() {
  Fnv1a64 hasher;
  for (std::int64_t z = -32; z < 32; ++z) {
    for (std::int64_t x = -32; x < 32; ++x) {
      hasher.write_i64(nexora::surface_height(0x0BEEF0BEEFULL, x, z));
    }
  }
  return hasher.finish();
}

std::uint64_t digest_chunk_cells(std::uint64_t* non_air_out, std::uint64_t* sections_out) {
  const nexora::ChunkShape shape;
  const nexora::GeneratedChunk chunk = nexora::generate_chunk(0x0BEEF0BEEFULL, shape, 0, 0);
  Fnv1a64 hasher;
  for (const auto& entry : chunk.sections) {
    hasher.write_i64(entry.first);
    const std::size_t volume = shape.volume();
    for (std::size_t index = 0; index < volume; ++index) {
      hasher.write_u32(entry.second.get_by_index(index));
    }
  }
  *non_air_out = chunk.non_air;
  *sections_out = chunk.sections.size();
  return hasher.finish();
}

// A floor with a wall at x == 4, three cells tall.
bool solid_fixture(std::int64_t x, std::int64_t y, std::int64_t z) {
  (void)z;
  return y < 0 || (x == 4 && y < 3);
}

std::uint64_t digest_sweep() {
  Fnv1a64 hasher;
  nexora::Rng driver(0x9E3779B97F4A7C15ULL);
  for (int i = 0; i < 4000; ++i) {
    const auto component = [&driver]() {
      return (static_cast<double>(driver.next_u64() & 0xFFFF) / 65535.0) * 8.0 - 4.0;
    };
    const double cx = component();
    const double cy = component();
    const double cz = component();
    nexora::Aabb box{{cx - 0.3, cy - 0.9, cz - 0.3}, {cx + 0.3, cy + 0.9, cz + 0.3}};
    for (int axis = 0; axis < 3; ++axis) {
      const double delta = component();
      const nexora::AxisSweep sweep = nexora::sweep_axis(solid_fixture, box, axis, delta);
      hasher.write_u64(static_cast<std::uint64_t>(sweep.blocked));
      std::uint64_t bits = 0;
      std::memcpy(&bits, &sweep.allowed, sizeof bits);
      hasher.write_u64(bits);
    }
  }
  // The exact case that produced the fall-through-the-floor defect: feet at
  // 1.9 - 0.9, which is 0.9999999999999999 rather than 1.0.
  const double feet = 1.9 - 0.9;
  nexora::Aabb standing{{0.2, feet, 0.2}, {0.8, feet + 1.8, 0.8}};
  const nexora::AxisSweep resting = nexora::sweep_axis(
      [](std::int64_t x, std::int64_t y, std::int64_t z) {
        (void)x;
        (void)z;
        return y < 1;
      },
      standing, 1, -0.0027);
  hasher.write_u64(static_cast<std::uint64_t>(resting.blocked));
  std::uint64_t bits = 0;
  std::memcpy(&bits, &resting.allowed, sizeof bits);
  hasher.write_u64(bits);
  return hasher.finish();
}

}  // namespace

int main() {
  emit("rng.splitmix64", digest_splitmix64());
  emit("rng.next_below", digest_next_below());
  emit("rng.stream_seeds", digest_stream_seeds());
  emit("rng.positional", digest_positional_rng());
  emit("hash.fnv1a", digest_fnv1a());
  emit("hash.crc32", digest_crc32());
  emit("spatial.section_of", digest_section_of());
  emit("world.surface_height", digest_surface_height());
  std::uint64_t non_air = 0;
  std::uint64_t sections = 0;
  const std::uint64_t cells = digest_chunk_cells(&non_air, &sections);
  emit("world.chunk_cells", cells);
  emit("world.chunk_non_air", non_air);
  emit("world.chunk_sections", sections);
  emit("physics.sweep", digest_sweep());
  return 0;
}

File: engine/benchmark/src/conformance.rs
//! Digests that prove a second stack computes the same thing.
//!
//! `NEXORA TECHNOLOGY BENCHMARK PLAN.md` rule 1 requires the **same logical
//! workload** across candidate stacks, and rule 5 forbids deciding from one.
//! Comparing two stacks therefore needs a step comparing *implementations*
//! before it compares *timings* — otherwise the honest reading of a fast
//! number is "the other one might just be doing less".
//!
//! `DEBT-0005` made that point within one language: the shift shortcut was
//! timed against `div_euclid` only after a test proved they agreed across
//! 20 000 coordinates. Across two languages the risk is far higher — Euclidean
//! division, rounding half away from zero, and unsigned-to-signed casts all
//! differ between Rust and C++ by default.
//!
//! Each function here folds a kernel's output into an FNV-1a 64 digest.
//! `benchmarks/cpp/src/conformance.cpp` emits the same names, and
//! `scripts/compare-stacks.sh` fails on any difference.

use nexora_foundation::error::Result;
use nexora_foundation::hashing::{crc32, fnv1a64, Fnv1a64};
use nexora_foundation::rng::{derive_stream_seed, positional_rng, Rng, SeedStream};
use nexora_foundation::spatial::{BlockPos, ChunkCoord, ChunkShape};
use nexora_foundation::time::CalendarConfig;
use nexora_physics::collision::sweep_axis;
use nexora_physics::math::{Aabb, Axis, Vec3};
use nexora_physics::voxel::{VoxelShape, VoxelSource};
use nexora_world::world::{World, WorldDescriptor};

/// The seed every digest that needs one uses.
const DIGEST_SEED: u64 = 0x0BEEF_0BEEF;

/// One named digest.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Digest {
    /// Kernel name, matching the C++ reference.
    pub name: &'static str,
    /// The folded value.
    pub value: u64,
}

/// Every digest, in the order the C++ reference emits them.
///
/// # Errors
///
/// Returns an error when the benchmark world cannot be built.
pub fn digests() -> Result<Vec<Digest>> {
    let (chunk_cells, chunk_non_air, chunk_sections) = chunk_digest()?;
    Ok(vec![
        Digest {
            name: "rng.splitmix64",
            value: splitmix64_digest(),
        },
        Digest {
            name: "rng.next_below",
            value: next_below_digest(),
        },
        Digest {
            name: "rng.stream_seeds",
            value: stream_seed_digest(),
        },
        Digest {
            name: "rng.positional",
            value: positional_digest(),
        },
        Digest {
            name: "hash.fnv1a",
            value: fnv1a_digest(),
        },
        Digest {
            name: "hash.crc32",
            value: crc32_digest(),
        },
        Digest {
            name: "spatial.section_of",
            value: section_of_digest(),
        },
        Digest {
            name: "world.surface_height",
            value: surface_height_digest()?,
        },
        Digest {
            name: "world.chunk_cells",
            value: chunk_cells,
        },
        Digest {
            name: "world.chunk_non_air",
            value: chunk_non_air,
        },
        Digest {
            name: "world.chunk_sections",
            value: chunk_sections,
        },
        Digest {
            name: "physics.sweep",
            value: sweep_digest(),
        },
    ])
}

fn splitmix64_digest() -> u64 {
    let mut rng = Rng::from_seed(DIGEST_SEED);
    let mut hasher = Fnv1a64::new();
    for _ in 0..64 {
        hasher.write_u64(rng.next_u64());
    }
    hasher.finish()
}

fn next_below_digest() -> u64 {
    let mut rng = Rng::from_seed(12_345);
    let mut hasher = Fnv1a64::new();
    for _ in 0..2_000 {
        hasher.write_u64(rng.next_below(25));
        hasher.write_u64(rng.next_below(1));
        hasher.write_u64(rng.next_below(0));
        hasher.write_u64(rng.next_below(u64::MAX));
    }
    hasher.finish()
}

fn stream_seed_digest() -> u64 {
    let mut hasher = Fnv1a64::new();
    let streams = [
        SeedStream::Terrain,
        SeedStream::Biome,
        SeedStream::Cave,
        SeedStream::Vegetation,
        SeedStream::Weather,
        SeedStream::Structure,
        SeedStream::Civilization,
        SeedStream::Ai,
        SeedStream::History,
        SeedStream::Loot,
    ];
    for seed in [0u64, 1, 0xDEAD_BEEF, u64::MAX] {
        for stream in streams {
            hasher.write_u64(derive_stream_seed(seed, stream));
        }
    }
    hasher.finish()
}

fn positional_digest() -> u64 {
    let mut hasher = Fnv1a64::new();
    let mut driver = Rng::from_seed(0xA5A5_A5A5);
    for _ in 0..512 {
        let x = (driver.next_u64() as i64) % 100_000;
        let y = (driver.next_u64() as i64) % 1_000;
        let z = (driver.next_u64() as i64) % 100_000;
        hasher.write_u64(positional_rng(0x5EED, SeedStream::Terrain, x, y, z).next_u64());
    }
    hasher.finish()
}

fn fnv1a_digest() -> u64 {
    let mut hasher = Fnv1a64::new();
    for input in [
        "",
        "a",
        "foobar",
        "nexora:block/stone",
        "the quick brown fox",
    ] {
        hasher.write_u64(fnv1a64(input.as_bytes()));
    }
    let mut prefixed = Fnv1a64::new();
    prefixed.write_str("ab");
    prefixed.write_str("c");
    hasher.write_u64(prefixed.finish());
    hasher.finish()
}

fn crc32_digest() -> u64 {
    let mut hasher = Fnv1a64::new();
    for input in ["", "123456789", "NEXORA", "nexora:save/world_header"] {
        hasher.write_u64(u64::from(crc32(input.as_bytes())));
    }
    let mut payload = vec![0u8; 4_096];
    let mut driver = Rng::from_seed(7);
    for byte in &mut payload {
        *byte = driver.next_u64() as u8;
    }
    hasher.write_u64(u64::from(crc32(&payload)));
    hasher.finish()
}

fn section_of_digest() -> u64 {
    let shape = ChunkShape::cubic_default();
    let mut hasher = Fnv1a64::new();
    let mut driver = Rng::from_seed(0x2545_F491_4F6C_DD1D);
    for _ in 0..20_000 {
        let x = (driver.next_u64() as i64) % 4_000_000;
        let y = (driver.next_u64() as i64) % 4_000_000;
        let z = (driver.next_u64() as i64) % 4_000_000;
        let coord = shape.section_of(BlockPos::new(x, y, z));
        hasher.write_u64(coord.x as u64);
        hasher.write_u64(coord.y as u64);
        hasher.write_u64(coord.z as u64);
    }
    for value in [-33i64, -32, -31, -1, 0, 1, 31, 32, 33] {
        let coord = shape.section_of(BlockPos::new(value, value, value));
        hasher.write_u64(coord.x as u64);
    }
    hasher.finish()
}

fn digest_world() -> Result<World> {
    let descriptor = WorldDescriptor::new("conformance", DIGEST_SEED)?;
    let mut world = World::create(descriptor, CalendarConfig::earthlike())?;
    world.bring_online()?;
    Ok(world)
}

fn surface_height_digest() -> Result<u64> {
    let world = digest_world()?;
    let mut hasher = Fnv1a64::new();
    for z in -32..32 {
        for x in -32..32 {
            hasher.write_u64(world.surface_height(x, z) as u64);
        }
    }
    Ok(hasher.finish())
}

fn chunk_digest() -> Result<(u64, u64, u64)> {
    let world = digest_world()?;
    let chunk = world.generate_chunk(ChunkCoord::new(0, 0))?;
    let shape = world.descriptor().shape;
    let volume = shape.volume();

    let mut hasher = Fnv1a64::new();
    let mut non_air = 0u64;
    let indices = chunk.section_indices();
    for section_y in &indices {
        hasher.write_u64(*section_y as u64);
        let Some(section) = chunk.section(*section_y) else {
            continue;
        };
        non_air += u64::from(section.non_air_count());
        for index in 0..volume {
            let local = shape.local_from_index(index)?;
            hasher.write_u64(u64::from(section.get(local)?.0));
        }
    }
    Ok((hasher.finish(), non_air, indices.len() as u64))
}

/// A floor with a wall at `x == 4`, three cells tall. Mirrors the C++ fixture.
/// A step of solid cells at `x == 4`, over solid ground below `y == 0`.
///
/// `pub` so the timing suites can sweep against the *same* fixture the
/// conformance digest verifies, and so the C++ reference — which mirrors this
/// predicate exactly — is timed against identical geometry. A benchmark pair
/// running on two different worlds compares two different problems.
pub struct SweepFixture;

impl VoxelSource for SweepFixture {
    fn shape_at(&self, position: BlockPos) -> VoxelShape {
        if position.y < 0 || (position.x == 4 && position.y < 3) {
            VoxelShape::SOLID
        } else {
            VoxelShape::Empty
        }
    }
}

/// A floor whose top face is `y = 1`, for the standing case below.
struct RaisedFloor;

impl VoxelSource for RaisedFloor {
    fn shape_at(&self, position: BlockPos) -> VoxelShape {
        if position.y < 1 {
            VoxelShape::SOLID
        } else {
            VoxelShape::Empty
        }
    }
}

fn sweep_digest() -> u64 {
    let mut hasher = Fnv1a64::new();
    let mut driver = Rng::from_seed(0x9E37_79B9_7F4A_7C15);
    // The arithmetic is written out rather than fused: `mul_add` would round
    // once where the C++ expression rounds twice, and the two stacks would stop
    // agreeing on the last bit.
    let component = |driver: &mut Rng| ((driver.next_u64() & 0xFFFF) as f64 / 65_535.0) * 8.0 - 4.0;
    for _ in 0..4_000 {
        let cx = component(&mut driver);
        let cy = component(&mut driver);
        let cz = component(&mut driver);
        let aabb = Aabb {
            min: Vec3::new(cx - 0.3, cy - 0.9, cz - 0.3),
            max: Vec3::new(cx + 0.3, cy + 0.9, cz + 0.3),
        };
        for axis in Axis::ALL {
            let delta = component(&mut driver);
            let sweep = sweep_axis(&SweepFixture, aabb, axis, delta);
            hasher.write_u64(u64::from(sweep.is_blocked()));
            hasher.write_u64(sweep.allowed.to_bits());
        }
    }

    // The exact case that produced the fall-through-the-floor defect.
    let feet = 1.9 - 0.9;
    let standing = Aabb {
        min: Vec3::new(0.2, feet, 0.2),
        max: Vec3::new(0.8, feet + 1.8, 0.8),
    };
    let resting = sweep_axis(&RaisedFloor, standing, Axis::Y, -0.0027);
    hasher.write_u64(u64::from(resting.is_blocked()));
    hasher.write_u64(resting.allowed.to_bits());
    hasher.finish()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn every_digest_has_a_distinct_name() {
        let digests = digests().expect("digests");
        let mut names: Vec<_> = digests.iter().map(|entry| entry.name).collect();
        names.sort_unstable();
        let before = names.len();
        names.dedup();
        assert_eq!(names.len(), before, "two digests share a name");
        assert!(before >= 12);
    }

    #[test]
    fn digests_are_stable_across_calls() {
        // If a digest depended on anything but its inputs, the comparison
        // against another stack would fail for a reason that has nothing to do
        // with that stack.
        let first = digests().expect("digests");
        for _ in 0..3 {
            assert_eq!(digests().expect("digests"), first);
        }
    }

    #[test]
    fn the_chunk_digest_describes_a_real_chunk() {
        let digests = digests().expect("digests");
        let find = |name: &str| {
            digests
                .iter()
                .find(|entry| entry.name == name)
                .expect("present")
                .value
        };
        assert!(
            find("world.chunk_sections") > 1,
            "no sections were generated"
        );
        assert!(
            find("world.chunk_non_air") > 1_000_000,
            "the chunk is empty"
        );
    }
}
